adlik_serving/runtime/enflame/model/enflame_model.cc

```diff
@@ -0,0 +1,298 @@
+  do {                                                  \
+    const ::tensorflow::Status _status = (__VA_ARGS__); \
+    if (TF_PREDICT_FALSE(!_status.ok()))                \
+      return cub::Status(_status.code());               \
+  } while (0)
+
+#define MY_RETURN_IF_ERROR_TF(...)                      \
+  do {                                                  \
+    const ::tensorflow::Status _status = (__VA_ARGS__); \
+    if (TF_PREDICT_FALSE(!_status.ok()))                \
+      return _status;                                   \
+  } while (0)
+
+struct Buffer {
+  Buffer(std::string name, tensorflow::DataType data_type, void* address = nullptr)
+      : name(name), data_type(data_type), address(address) {
+  }
+  ~Buffer() {
+    TopsInference::mem_free(address);
+  }
+  std::string name;
+  tensorflow::DataType data_type;
+  size_t byte_1batch_size;
+  size_t byte_size;
+  std::vector<int> shape;
+  void* address = nullptr;
+};
+
+struct Instance : BatchProcessor {
+  Instance(const std::string& name, const adlik::serving::ModelConfigProto& config, TopsInference::IEngine* model)
+      : name(name), config(config), engine(model), stream(stream), max_batch_size(config.max_batch_size()) {
+  }
+  ~Instance() {
+    TopsInference::destroy_stream(stream);
+    TopsInference::release_engine(engine);
+    TopsInference::topsInference_finish();
+    INFO_LOG << "~Instance()";
+  };
+  tensorflow::Status init();
+
+private:
+  const std::string name;
+  const adlik::serving::ModelConfigProto config;
+  const int max_batch_size;
+  TopsInference::IEngine* engine;
+  TopsInference::topsInferStream_t stream;
+  std::vector<Buffer> inputs;
+  std::vector<Buffer> outputs;
+
+  tensorflow::Status processBatch(Batch<BatchingMessageTask>&) override;
+  tensorflow::Status mergeInputs(Batch<BatchingMessageTask>&);
+  tensorflow::Status splitOutputs(Batch<BatchingMessageTask>&);
+};
+
+tensorflow::Status Instance::init() {
+  auto init_buffer = [&](Buffer& buffer, const adlik::serving::DimsList& dims) {
+    buffer.shape.push_back(max_batch_size);
+    for (auto dim : dims)
+      buffer.shape.push_back(dim);
+    size_t typesize = GetDataTypeSize(buffer.data_type);
+    if (typesize <= 0)
+      return tensorflow::errors::InvalidArgument("unknow data_type ", buffer.data_type);
+    buffer.byte_1batch_size =
+        std::accumulate(++buffer.shape.begin(), buffer.shape.end(), 1, std::multiplies<int>()) * typesize;
+    buffer.byte_size = buffer.byte_1batch_size * max_batch_size;
+    return tensorflow::Status::OK();
+  };
+
+  // init inputs
+  for (int i = 0; i < config.input_size(); ++i) {
+    auto model_input = config.input(i);
+    inputs.emplace_back(model_input.name(), model_input.data_type());
+    MY_RETURN_IF_ERROR_TF(init_buffer(inputs.back(), model_input.dims()));
+    void* storage_on_device = inputs.back().address;
+    TopsInference::mem_alloc(&storage_on_device, inputs.back().byte_size);
```

Comment:
The allocated memory is asigned to storage_on_device, but inputs.back().address is still nullptr.